pkg/healthcheck/healthcheck_output.go

```diff
@@ -48,20 +52,142 @@ func (cr CheckResults) RunChecks(observer CheckObserver) bool {
+	}
+}
+
+// RunExtensionsChecks runs checks for each extension name passed into the `extensions` parameter
+// and handles formatting the output for each extension's check. This function also handles
+// finding the extension in the user's path and runs it.
+func RunExtensionsChecks(wout io.Writer, werr io.Writer, extensions []string, flags []string, output string) bool {
+	if output != JSONOutput {
+		headerTxt := "Linkerd extensions checks"
+		fmt.Fprintln(wout)
+		fmt.Fprintln(wout, headerTxt)
+		fmt.Fprintln(wout, strings.Repeat("=", len(headerTxt)))
+	}
+
+	spin := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+	spin.Writer = wout
+
+	success := true
+	for i, extension := range extensions {
+		var path string
+		args := append([]string{"check"}, flags...)
+		var err error
+		results := CheckResults{
+			Results: []CheckResult{},
+		}
+		extensionCmd := fmt.Sprintf("linkerd-%s", extension)
+
+		switch extension {
+		case "jaeger":
+			path = os.Args[0]
+			args = append([]string{"jaeger"}, args...)
+		case "viz":
+			path = os.Args[0]
+			args = append([]string{"viz"}, args...)
+		case "multicluster":
+			path = os.Args[0]
+			args = append([]string{"multicluster"}, args...)
+		default:
+			path, err = exec.LookPath(extensionCmd)
+			results.Results = []CheckResult{
+				{
+					Category:    CategoryID(extensionCmd),
+					Description: fmt.Sprintf("Linkerd extension command %s exists", extensionCmd),
+					Err:         err,
+					HintURL:     DefaultHintBaseURL + "extensions",
+					Warning:     true,
+				},
+			}
+		}
+
+		if err == nil {
+			if isatty.IsTerminal(os.Stdout.Fd()) {
+				spin.Suffix = fmt.Sprintf(" Running %s extension check", extension)
+				spin.Color("bold") // this calls spin.Restart()
+			}
+			plugin := exec.Command(path, args...)
+			var stdout, stderr bytes.Buffer
+			plugin.Stdout = &stdout
+			plugin.Stderr = &stderr
+			plugin.Run()
+			extensionResults, err := parseJSONCheckOutput(stdout.Bytes())
+			spin.Stop()
+			if err != nil {
+				command := fmt.Sprintf("%s %s", path, strings.Join(args, " "))
+				if len(stderr.String()) > 0 {
+					err = errors.New(stderr.String())
+				} else {
+					err = fmt.Errorf("invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%s]", command, stdout.String(), err)
+				}
+				results.Results = append(results.Results, CheckResult{
+					Category:    CategoryID(extensionCmd),
+					Description: fmt.Sprintf("Running: %s", command),
+					Err:         err,
+					HintURL:     DefaultHintBaseURL + "extensions",
+				})
+				success = false
+			} else {
+				results.Results = append(results.Results, extensionResults.Results...)
+			}
+		}
+
+		var containsErr bool
+		for _, res := range results.Results {
+			if res.Err != nil {
+				containsErr = true
+				break
+			}
+		}
+
+		if output != JSONOutput && i < len(extensions) {
+			// add a new line to space out each check output
+			fmt.Fprintln(wout)
+			if output == ShortOutput && !containsErr {
```

Comment:
This code is to print the header for each Linkerd extension, right?  Why only print the header in short mode?  Why only print the header if there are no errors?  I'm not sure I understand the intent here.

Reply:
This is definitely a bit hacky but the intent here is to display a header for the check we are running in the case where there are no errors or warnings. For each check category, we do not show its header when everything is all green. When running this in extension checks, the output looks a bit weird when everything is green and we don't have this condition in place i.e. 

```
Linkerd core checks
===================
...

Linkerd extensions checks
=========================

Status check results are √
OK: 1, WARNING: 0, FAIL: 0

Status check results are √
OK: 11, WARNING: 0, FAIL: 0
```

It isn't clear which extension check was ran, so having the header being printed, in short mode when all extensions checks pass shows us which checks were ran.

If we flip it and always print a header, then we'll end up duplicating the header section when we are running extension checks under normal table mode:
```
Linkerd core checks
===================

...

Linkerd extensions checks
=========================

linkerd-multicluster
--------------------
linkerd-multicluster
--------------------
√ Link CRD exists

Status check results are √

linkerd-viz
-----------
linkerd-viz
-----------
√ linkerd-viz Namespace exists
√ linkerd-viz ClusterRoles exist
√ linkerd-viz ClusterRoleBindings exist
√ tap API server has valid cert
√ tap API server cert is valid for at least 60 days
√ tap API service is running
√ linkerd-viz pods are injected
√ viz extension pods are running
√ prometheus is installed and configured correctly
√ can initialize the client
√ viz extension self-check

Status check results are √
```

I can add a comment above this section if it isn't immediately clear why we need this.

Reply:
So it sounds like we're actually adding two new printing styles in short mode:
* When rendering the main checks, we want to only print errors and only print category headers if that category has an error in it
* When rendering the extension checks, we want to only print errors but always print the category header

Rather than distributing this logic between here and  prettyPrintResults, I think we need multiple different `prettyPrintResults` functions that perform each of the correct behaviors.  e.g. `prettyPrintShortResults` and `prettyPrintShortExtensionResults`.  Each of these pretty print functions will have a lot in common (for example, displaying a spinner during retries, printing the category header, etc.) and can be built up from shared helper functions to avoid excessive duplication.